src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java

```diff
@@ -615,32 +578,96 @@ static void checkApiRateLimit(@NonNull TaskListener listener, GitHub github)
-                    Long lastUse = entry.getValue();
-                    if (lastUse == null || lastUse < threshold) {
-                        iterator.remove();
-                        unused(entry.getKey());
+            synchronized (connections) {
+                GitHubConnection.removeAllUnused(threshold);
+            }
+        }
+    }
+
+    private static class GitHubConnection {
+        @NonNull
+        private final GitHub gitHub;
+
+        @CheckForNull
+        private final Cache cache;
+
+        private final boolean cleanupCacheFolder;
+        private int usageCount = 1;
+        private Long lastUsed = System.currentTimeMillis();
+
+        private GitHubConnection(GitHub gitHub, Cache cache, boolean cleanupCacheFolder) {
+            this.gitHub = gitHub;
+            this.cache = cache;
+            this.cleanupCacheFolder = cleanupCacheFolder;
+        }
+
+        /**
+         * Gets the {@link GitHub} instance for this connection
+         *
+         * @return the {@link GitHub} instance
+         */
+        public GitHub getGitHub() {
+            return gitHub;
+        }
+
+        @CheckForNull
+        private static GitHubConnection lookup(@NonNull ConnectionId connectionId) {
+            GitHubConnection record;
+            record = connections.get(connectionId);
+            if (record != null) {
+                record.usageCount += 1;
+                record.lastUsed = System.currentTimeMillis();
+            }
+            return record;
+        }
+
+        @NonNull
+        private static GitHubConnection connect(@NonNull ConnectionId connectionId, @NonNull GitHub gitHub, @CheckForNull Cache cache, boolean cleanupCacheFolder) {
+            GitHubConnection record = new GitHubConnection(gitHub, cache, cleanupCacheFolder);
+            connections.put(connectionId, record);
+            reverseLookup.put(record.gitHub, record);
+            return record;
+        }
+
+
+        private void release() {
+            if (this.usageCount <= 1) {
+                this.usageCount = 0;
+                this.lastUsed = System.currentTimeMillis();
+            } else {
+                this.usageCount -= 1;
+            }
+        }
+
+        private static void removeAllUnused(long threshold) throws IOException {
+            for (Iterator<Map.Entry<ConnectionId, GitHubConnection>> iterator = connections.entrySet().iterator();
+                 iterator.hasNext(); ) {
+                GitHubConnection record = Objects.requireNonNull(iterator.next().getValue());
+                Long lastUse = record.lastUsed;
+                if (record.usageCount == 0 && lastUse < threshold) {
+                    iterator.remove();
+                    reverseLookup.remove(record.gitHub);
+                    if (record.cache != null && record.cleanupCacheFolder) {
+                        record.cache.delete();
+                        record.cache.close();
```

Comment:
Should we catch `IOException` here so that even if there is a problem with one connection we still try to clean up the rest? (We could still throw it at the end of the method; just add suppressed exceptions if more than one cache has an issue)

Reply:
+1 for cleanup